packages-exp/auth-exp/src/core/user/token_manager.test.ts

```diff
@@ -187,6 +188,97 @@ describe('core/user/token_manager', () => {
       const tokens = (await stsTokenManager.getToken(auth))!;
       expect(tokens).to.eql('token');
     });
+
+    describe('in passthrough mode', () => {
+      const fakeCustomToken = 'fake-custom-token';
+      const provider = {
+        async getCustomToken(): Promise<string> {
+          return fakeCustomToken;
+        }
+      };
+
+      it('refreshes the token if forceRefresh is true and custom token provider is set', async () => {
+        Object.assign(stsTokenManager, {
+          accessToken: 'old-access-token',
+          expirationTime: now + 100_000,
+          isPassthroughMode: true
+        });
+        setCustomTokenProvider(auth, provider);
+        sinon
+          .stub(auth, '_refreshWithCustomTokenProvider')
+          .returns(Promise.resolve('new-access-token'));
+
+        const tokens = (await stsTokenManager.getToken(auth, true))!;
+
+        expect(tokens).to.eq('new-access-token');
+      });
+
+      it('throws an error if forceRefresh is true and custom token provider is not set', async () => {
+        Object.assign(stsTokenManager, {
+          accessToken: 'token',
+          expirationTime: now + 100_000,
+          isPassthroughMode: true
+        });
+
+        await expect(stsTokenManager.getToken(auth, true)).to.be.rejectedWith(
+          FirebaseError,
+          'auth/internal-error'
+        );
+      });
+
+      it('returns access token during buffer window if custom token provider is not set', async () => {
+        Object.assign(stsTokenManager, {
+          accessToken: 'token',
+          expirationTime: now + Buffer.TOKEN_REFRESH,
+          isPassthroughMode: true
+        });
+
+        const tokens = (await stsTokenManager.getToken(auth))!;
+
+        expect(tokens).to.eq('token');
+      });
+
+      it('refreshes the token during buffer window if custom token provider is set', async () => {
+        Object.assign(stsTokenManager, {
+          accessToken: 'old-access-token',
+          expirationTime: now + Buffer.TOKEN_REFRESH / 2,
+          isPassthroughMode: true
+        });
+        setCustomTokenProvider(auth, provider);
+        sinon
+          .stub(auth, '_refreshWithCustomTokenProvider')
+          .returns(Promise.resolve('new-access-token'));
+
+        const tokens = (await stsTokenManager.getToken(auth))!;
+
+        expect(tokens).to.eq('new-access-token');
+      });
+
+      it('throws an error if the token is expired beyond the buffer window', async () => {
```

Comment:
Can you also add a test for when the token is expired, but there is a `CustomTokenProvider` set (unless you already have this test case and I just missed it)?

Reply:
Sick, added!